client/my-sites/themes/themes-site-selector-modal.jsx

```diff
@@ -14,49 +17,98 @@ import { trackClick } from './helpers';
 
 const ThemesSiteSelectorModal = React.createClass( {
 	propTypes: {
-		name: React.PropTypes.string.isRequired,
-		label: React.PropTypes.string.isRequired,
-		header: React.PropTypes.string.isRequired,
-		selectedTheme: PropTypes.object.isRequired,
-		onHide: PropTypes.func,
-		action: PropTypes.func,
+		options: React.PropTypes.objectOf( React.PropTypes.shape( {
+			label: React.PropTypes.string,
+			header: React.PropTypes.string,
+			action: React.PropTypes.func
+		} ) ),
+		selectedSite: React.PropTypes.object,
 		// Will be prepended to site slug for a redirect on selection
 		sourcePath: PropTypes.string.isRequired,
 	},
 
+	getInitialState() {
+		return {
+			selectedTheme: {},
+			selectedOption: {},
+		};
+	},
+
 	redirectAndCallAction( site ) {
+		const action = this.state.selectedOption.action;
+		const theme = this.state.selectedTheme;
 		/**
 		 * Since this implies a route change, defer it in case other state
 		 * changes are enqueued, e.g. setSelectedTheme.
 		 */
 		defer( () => {
 			trackClick( 'site selector', this.props.name );
 			page( this.props.sourcePath + '/' + site.slug );
-			this.props.action( this.props.selectedTheme, site );
+			action( theme, site );
 		} );
 	},
 
+	showSiteSelectorModal( option, theme ) {
+		this.setState( { selectedTheme: theme, selectedOption: option } );
+	},
+
+	hideSiteSelectorModal() {
+		this.showSiteSelectorModal( {}, {} );
+	},
+
+	isThemeOrActionSet() {
+		return ! isEmpty( this.state.selectedTheme ) || ! isEmpty( this.state.selectedOption );
+	},
+
 	render() {
+		const { defaultOption } = this.props;
+		const wrappedDefaultOption = Object.assign(
+			{},
+			defaultOption,
+			defaultOption.action
+				? { action: theme => this.showSiteSelectorModal( defaultOption, theme ) }
```

Comment:
`wrappedOptions` and `wrappedDefaultOption` might be able to share some code here? (fine to iterate in another PR)